language-server/modules/langserver-core/src/main/java/org/ballerinalang/langserver/completions/providers/context/util/ServiceTemplateGenerator.java

```diff
@@ -0,0 +1,530 @@
+                generateServiceSnippet(item, shouldImport, ctx)).collect(Collectors.toList());
+    }
+
+    /**
+     * Generates and returns the service templates for a given completion context.
+     *
+     * @param ctx Completion context.
+     * @return {@link List<LSCompletionItem>} List of completion items.
+     */
+    public List<LSCompletionItem> getServiceTemplates(BallerinaCompletionContext ctx) {
+        List<LSCompletionItem> completionItems = new ArrayList<>();
+        Set<String> processedModuleList = new HashSet<>();
+
+        //Find listeners from current imports and generate completion items.
+        Map<ImportDeclarationNode, ModuleSymbol> currentDocImports = ctx.currentDocImportsMap();
+        currentDocImports.forEach((importNode, moduleSymbol) -> {
+            String orgName = importNode.orgName().isEmpty() ? "" : importNode.orgName().get().orgName().text();
+            String moduleName = importNode.moduleName().stream()
+                    .map(Token::text)
+                    .collect(Collectors.joining("."));
+            String modulePrefix;
+            if (importNode.prefix().isEmpty()) {
+                modulePrefix = CommonUtil.escapeReservedKeyword(
+                        importNode.moduleName().get(importNode.moduleName().size() - 1).text());
+            } else {
+                modulePrefix = CommonUtil.escapeReservedKeyword(importNode.prefix().get().prefix().text());
+            }
+            String moduleHash = generateModuleHash(orgName, moduleName);
+            if (processedModuleList.contains(moduleHash)) {
+                return;
+            }
+
+            Pair<String, String> key = Pair.of(moduleName, orgName);
+            //check if the module has already been processed to the cache.
+            if (this.moduleListenerMetaDataMap.containsKey(key)) {
+                moduleListenerMetaDataMap.get(key).forEach(item ->
+                        completionItems.add(generateServiceSnippet(item, false, ctx)));
+                processedModuleList.add(moduleHash);
+                return;
+            }
+
+            //Check if the module belongs to the current project. 
+            //If it is not from the current project populate to the cache.
+            if (!getModuleNamesOfCurrentProject(ctx).contains(moduleHash)) {
+                completionItems.addAll(generateAndPopulate(moduleSymbol, false, ctx));
+                processedModuleList.add(moduleHash);
+                return;
+            }
+
+            moduleSymbol.allSymbols().stream().filter(listenerPredicate())
+                    .forEach(listener -> generateServiceSnippetMetaData(listener, orgName, moduleName, modulePrefix)
+                            .ifPresent(item -> completionItems.add(generateServiceSnippet(item, false, ctx))));
+            processedModuleList.add(moduleHash);
+        });
+
+        //Generate service templates for listeners from the distribution
+        if (this.isInitialized) {
```

Comment:
Since we have a public method for checking the initialization status, let's use that. That will ensure, even in the future, if do any thread safe logic improvements that also get affected for all the usages